### ISSUE TITLE
Allow additional scopes with own oauth

### ISSUE THREAD

**Issue by chrisherold** [r-dbi/bigrquery#297]
I read #132 where this was discussed a couple years ago, as I have a number of tables in BigQuery that depend on underlying Google Sheets, and understand why opening up the bigrquery oauth app to additional scopes could be problematic.

But with the addition of #130 to allow users to bring their own oauth app, a user could control scopes within their own app (ability to monitor logs, revoke tokens, delete the oauth app). One could edit the `set_oauth2.0_cred` function like:
 ```
function(app = NULL, custom_scopes = NULL) {
  bq_scopes <- ifelse(is.null(app) | is.null(custom_scopes), bq_scopes, custom_scopes)
  cred <- oauth2.0_token(bq_endpoint, app %||% bq_app, scope = bq_scopes)
  set_access_cred(cred)
}
```
where `bq_scopes` would remain untouched if either `app` or `custom_scopes` is not provided to the function.

Currently it's possible to workaround this by doing the following:
```
library("httr")
library("bigrquery")

custom_scopes <- c(
  "https://www.googleapis.com/auth/bigquery",
  "https://www.googleapis.com/auth/cloud-platform",
  "https://www.googleapis.com/auth/drive.readonly"
)
 
assignInNamespace("bq_scopes",custom_scopes,"bigrquery")

my_app <- httr::oauth_app("google","[client_id]","[client_secret]")

bigrquery::set_oauth2.0_cred(my_app)
```
But doing `assignInNamespace` doesn't seem to be the best solution for the long run.

Does the ability to add custom scopes when using a custom oauth app make sense here?

**Comment by hadley** [comment]
@jennybc I think this is a question for you

**Comment by jennybc** [comment]
Background: scopes aren't persistently associated with an oauth app but are combined with the app at the moment of token request. So I'd rephrase your request as wanting control over scopes. Whether with the built-in app or your own.

In #292, I'm working to swap out bigrquery's auth system to a package called gargle that can be used across all Google APIs. Any package that uses gargle will have a function like `bq_auth()` to get a token and it has a `scopes` argument, with reasonable defaults. So, this is coming.

``` r
bq_auth <- function(email = NULL,
                     path = NULL,
                     scopes = c(
                       "https://www.googleapis.com/auth/bigquery",
                       "https://www.googleapis.com/auth/cloud-platform"
                     ),
                     cache = getOption("gargle.oauth_cache"),
                     use_oob = getOption("gargle.oob_default")) { ... }
```

**Comment by jennybc** [comment]
This is available now, in the dev versions of bigrquery (and google drive). To continue with your example, try this:

``` r
# devtools::install_github("r-dbi/bigrquery")
# devtools::install_github("tidyverse/googledrive")

library("bigrquery")
library("googledrive")

# Assumption: the OAuth2 creds associated with this client ID and secret are
# from a GCP project with both the BigQuery and Drive APIs enabled
bq_auth_configure(
  app = httr::oauth_app("google","[client_id]","[client_secret]")
)

custom_scopes <- c(
  "https://www.googleapis.com/auth/bigquery",
  "https://www.googleapis.com/auth/cloud-platform",
  "https://www.googleapis.com/auth/drive.readonly"
)

bq_auth(scopes = custom_scopes)
drive_auth(token = bq_token())

# now work with bigrquery and googledrive as usual
# or retrieve the token via bq_token() or drive_token() to use in
# httr requests, more generally
```

I'm closing this because I have successfully done similar with Drive and Sheets. But if you are willing to install the dev versions and verify the above does exactly what you want with BigQuery and Sheets/Drive, please do come back and confirm all is working.
